### DISCUSSION TITLE
Github Variable referenced within a Github Secret

### DISCUSSION THREAD

**Discussion by arron-paul** [community/community#86569] [Actions] [answered]
### Select Topic Area

Question

### Body

I have a query regarding the interoperability of GitHub Secrets and GitHub Variables in workflows.

Is it possible to reference GitHub Variables such as `${{ github.ref_name }}` or `${{ vars.deploy_path }}` from directly within a GitHub Secret?

The scenario I'm considering involves creating a GitHub Secret to store the destination path for transferring built artifacts.

As an illustration, consider the following example of a GitHub Secret containing a destination path: `/var/www/projects/my-project/${{ github.ref_name }}/public-html` which would evaluate to `/var/www/projects/my-project/main/public-html` when performing a `push` on the `main` branch.

**Comment by ChristopherHX** [comment, marked as answer]
> Is it possible to reference GitHub Variables such as `${{ github.ref_name }}` or `${{ vars.deploy_path }}` from directly within a GitHub Secret?

As far I know, no this is not possible.

I can imagine you create a secret `X` with value `/var/www/projects/my-project/{0}/public-html` (escape `{` with `{{`, and `}` with `}}`).
then use it like `${{ format(secrets.X, github.ref_name) }}`.
One problem with this could be that your secret get's exposed to the log by doing it like that.. (Because you replace parts of the secret).

To get around the masking problem, you could create a reusable workflow defining secret `X` in it's on trigger.
Then you could rewrite your raw secret to contain data from the GitHub Context.
- set the secret `X` agan to value`/var/www/projects/my-project/{0}/public-html`
- ```yaml
  jobs:
    call:
      uses: .... # path to reusable workflow
      secrets:
        X: ${{ format(secrets.X, github.ref_name) }} # Warning using vars here had led to empty values here, even while vars is synactically allowed it didn't work at least for a long time
  ```

Otherwise you could create a script, which manipulates your secret to incoperate those values. Then use the `::add-mask::` command / `core.setSecret` js api, before exporting it to the env via `GITHUB_ENV` file command / `core.setVarable(...)` js api.

**Comment by arron-paul** [comment]
Thanks for the response, those solutions worked.

As my workflow is relatively simple at the moment, I've used the `format` method with a variable instead of a secret, as the path doesn't contain anything too sensitive. I noted the warning regarding the usage of `vars` in combination with `format` but it worked successfully in my case.

```
- name: Upload to remote server
  uses: appleboy/scp-action@master
  with:
    host: ${{ secrets.host }}
    username: ${{ secrets.user }}
    key: ${{ secrets.key }}
    source: "./"
    target: ${{ format(vars.path, github.ref_name) }}
```

**Reply by ChristopherHX** [reply]
Some bugs in gha are pretty special, where you decided to use the format expression access to `vars` usually worked without issues.
> I noted the warning regarding the usage of `vars` in combination with `format` but it worked successfully in my case.

Only and only if you would use `vars` in a `jobs.*.secrets.*` expression, they might have empty values.
For example `vars` should work as well in `jobs.*.with.*`, bugs are weird...
